### ISSUE TITLE
ToDo: followsups / extras / ideas [cont.]

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/TorZillaPrint#46]
New clean list after closing #28 

🔻 **general / perf**

* [x] added: debugging table: default collapsed
   - section perf: esp handy on android: only useful for dev with RFP timing disabled
   - added tor detection points
* [x] added: variable times (FF vs TB/Android) allocated to detect some failures: to still allow a speedy result, but also take latency into account
   - `domrect` (iframes blocked)
   - `woff` (pref disabled, downloadable fonts blocked)
* [x] split css @media from index.css (loaded first) into a separate file (media.css) which is loaded last with all the other css files (which are also all @media rules)

march 8th
* [x] lots of code simplification, cleanup, better/consistent naming: shrank size of js by about 10% despite adding more features etc
* [x] fields not coded for now have a placeholder text in faint dark grey saying they are not coded yet to avoid user confusion
* [x] improved/cleaner/better-named? globals.js
   - e.g. common results: start with `z`
* [x] improved dev debugging shit
   - [x] logExtra = show extras things such as
      - order of when zoom/screen/dpi etc run on various events/functions/button clicks
      - differences between unicode glyphs measureText vs clientrects/boundingClientRect
   - [x] logPerf (this is not section perf, but specific functions: e.g: getting a zoom value)
      - now uses a function to parse the input and output the timing: nicely aligned
   - [x] logStorage
      - pretty basic: but now it outputs nicely aligned results
      - will expand on this as I finish storage tests and add a zombie section
   - [x] runS <-- short for run simulation
      - work in progress: more to add
      - all simulated results have `[simulated]` appended
      - when true, it simulates what happens when things change/fail and if I can pick up on them: e.g **some examples**
      - e.g: new collation/error/math hashes not seen before
      - e.g. `isOS` is not set [this is in the widget test and is set once] : a lot of tests rely on `isOS`
      - e.g. the chrome or math os output doesn't match the widget one
      - e.g. a new branding size not seen on TB or Firefox before
      - e.g. ac-outputLatency fails
   - [x] runS [note]
      - because of all the variables involved, often there is more than one possible failure/change to pick up on, so you can re-run the section to output the alternative failure: e.g.
      - run1: `[math] os	Windows [!= widget] [simulated]`
      - run2: `[math] os	I haven't seen this Firefox math combo before [simulated]`
* [ ] runUA short for run UA simulation 🚀 
   - coming soon to a Firefox browser near you 👀 watch this space
   - this is almost ready to go, but I didn't upload it in the last batch: it detects any non RFP UA bullshit and lies and calls you out as high fingerprintable, if not outright unique :)
   - it counts the lies (one per ua navigator result) and points each one out for a BS meter rating

🔻 **screen**
* [x] added red/green LB notation for inner window (and orange LB if zoom != 100%) using the correct width and height steps of `50`, `100` or `200` pixels depending on the measurement
* [x] added red/green RFP New Window notation (and incorporated into the above LB orange notation if zoom !-100%) using width `200`s and height `100`s capped at `1000`
* [x] pure css field titles now show that they use `min-` and `min-device-` : end users are hopefully less confused now (when pixels results are 1 less than window and screen properties in JS)
* [x] matchMedia results are no longer rounded down when needed (to match pure css). This can do two things
   - when it shows a decimal, this indicates to the end user that there is entropy (but you would get that anyway, as the pure css is already 1 pixel less). This still doesn't let users know why it's 1 pixel less when rounding is not needed - but they should learn to compare like method with like method - it's clearly labelled as a `matchMedia` test
   - the decimal value may provide entropy based on the real DPI of the device
* [x] new window test
   - outputs a history of changes (if applicable) for both inner and outer measurements
   - this detects
      * bookmark toolbar padding (if personalbar is not allowed to be controlled = pref, e.g in Firefox)
      * LB borders being added (2px width, 1px height)
      * TB's new window inner/outer clamping (10 x 10) - and thus it bypasses the clamping
      * TB's outer window clamping (0,0) if RFP is disabled
      * the LBing step itself
   - it also notates how many checks it took for that change happened
* [x] added matchMedia `dppx` + `dpcm`
* [x] added matchMedia `display-mode`
* [x] added matchMedia `color`
* [x] added matchMedia `aspect-ratio` + `display-aspect-ratio` + `-moz-device-orientation`
* [x] added mozOrientation and orientation.angle
* [x] added RFP notation to orientation and to devicePixelRatio
* [x] added matchMedia `-webkit-` and `-moz-` device-pixel-ratio + RFP notation
* [x] resize event (desktop) updates devicePixelRatio + all orientation + display-mode + dpi/dppx/dpcm
   * note: almost everything now in screen section is now real-time
* [x] added `matches` notation for screen metrics (inner=outer=available=screen)
* [x] added `coordinates` notation for x/y screen metrics (must all be 0)
* [x] merged pixel depth and color depth and added RFP notation
* [x] added RFP notation to color
* [x] reduced lines by merging items into one line
   - css e.g. orientation + aspect-ratio + device-aspect-ratio
   - or matchMedia: e.g. orientation + aspect-ratio + device-aspect-ratio
   - or css and matchMedia into one line: e.g. display-mode
* [x] added new device pixel ratio PoC that uses border sizes to detect subpixels
   * [ ] ^^ update to use clientrects
* [ ] android: keyboard PoC, if the keyboard is a little slow to appear e.g. on first appearance after opening Firefox, then I can end up with 0 pixels: instead use a setInterval and keep checking for a static change
* [ ] android: keyboard height can be calculated wrong (by the toolbar height), and the toolbar height never found
   - not sure why this is: seems to be when there's lots of latency - e.g TB for Android
   - e.g. I get the screen measurements with toolbar hidden
   - this affects the firstwindow* global vars: that is I get the height with toolbar hidden and think it's the height with toolbar visible
   - I think what I need to do is record a history of changes and use logic to get the right values

march 8th
* [x] acat's devicePixelRatio leaks moved to extras page because
   - I already have another PoC that doesn't require user action and runs immediately on load
   - acat's PoCs are a bit buggy, require scrolling, and the results are affected by zoom/pinch-to-zoom

🔻 **ua**
* [x] added branding css image (it's a chrome:// svg file ) for desktop
   - I can already detect version, and people should always be up to date: but rollouts and uptake take time: so there is some value in this entropy
   - [x] allows me to refine `[resource://] browser` to
      - Firefox `[Stable]` or Firefox `[Developer or Nightly]`
      - tag Tor Browser  with `[alpha]` - am reluctant to tag anything with `[Stable]` in TB yet
   - tests done with
      - stable ESR/FF60 to current FF71, Dev 72 and Nightly 73
      - TB ESR60/68 stable, TB alpha 9.5a3
   - [x] ToDo: test older versions of FF Dev (Nightly should be the same) and TB alphas
      - **beta**: tested one random one from each release cycle from FF60 thru to FF72
      - **stable releases**: tested all from FF60-71, ESR60 and ESR68
      - **stable releases**: tested most of them on Linux (in case they differed: why would they, they don't btw)
      - **developer editions** : tested all from FF60-72
      - **nightly**: tested a 68 nightly in a VM and they seem to use the 270x48 one TB alpha now uses
      - tested all TB stable releases based on ESR60 + ESR68
      - spot tested a dozen TB alpha releases based on ESR60 + ESR68
   - [x] entropy output
      - cleaned up and expanded based on results
      - added element dimensions to the output
      - element dimensions output even if I don't know the "brand"
* [x] Tor Browser for Android: yay! finally found a chrome:// tor icon to output
* [x] fixed TB ESR60 version detection: my test 73 (added 10 days ago) caused it to fail
* [x] removed two items from the `chrome os` check as they only apply to WIndows XP/Vista which were dropped after FF52.9 - saves a few ms
* [x] get_browser_errors: removed the "InternalError: function nested too deeply" test
   - it's not needed, and was only included so there was an "InternalError" type
   - it took almost all of the get_browser_errors runtime
* [x] added a browser collation test
   - I sort 34 strings in 35 locales for 35 unique hashes, then hash those together
   - this overall hash is unique to Firefox and has been unchanged since at least FF60
   - you can see what I mean and have a play [here](https://thorin-oakenpants.github.io/testing/collation.html)

march 8th
* [x] collation test stripped down to six chars and six locales not in chrome, three of which provide entropy between Firefox releases. This collation versioning is shown in the output along with the locale differences
* [x] css line height: now handles when `getBoundingClientRect` is blocked
  - when getComputedStyle returns "normal" (since around FF72 from memory) the code falls back to clientrect: if blocked this caused screen.js to fail
* [x] navigator UA: now has a hash
* [x] navigator UA: now has a window.open hash check
* [x] ua feature detection now has a hash
   - this is made from 5 results that can't be changed/influenced 9e.g extentions/dpi/zooming etc
* [x] removed the generic `[feature detection] browser` output: the fact that other fields like error, chrome etc output already shows we know you're Firefox, so it's pretty much redundant
* [x] isFF global var being set is more robust, now uses a method that cannot (or shouldn't be able to) be affected
* [x] isFF is double checked and set if needed by the error strings function
* [x] errors  have been reduced to 5 results (4 on nightly or 75+: the fifth error is only shown if the error is thrown)
* [x] the five error strings (which cannot be altered by end users/extensions) allow a lot of version detection: currently 60-67, 68-69, 70, 71, 72-74, 75+: and based on when some features (which are not behind any prefs) roll out to dev, will automatically create more version detection. This version detection is shown in the output
* [x] version: reversed order to highest->lowest and skip subsequent tests once found
   - only three checks max to find nightly/dev/stable
   - only ten (or 12 until end of life) checks max to find current ESR
   - 2020: rapid release moves to 4 weeks: ESR  = every 13 releases
* [x] version: the error versioning is used to bypass/skip tests in the version feature detection: and each "version section function" still uses reverse order
  - e.g. if the error hash indicated 68-69: i only need test for 69, otherwise it is 68
  - e.g. if the error hash indicated 70: I don't need test for anything
  - this actually helps isolate false positives in the individual version tests
  - ^^ case in point: Tor Browser ESR60 on windows actually passes the 73 test
  - ^^ I now output that fact on TB ESR60 as a difference between ESR60 vs TB (not that we need it: we already know they're Tor Browser from other fingerprinted items)
  - overall some good perf increases: versioning is now usually 1 to 3 ms, even for old releases
* navigator UA now includes vendor and vendorSub - they only show if they have a value
   - needed for output/hashing etc with UA spoofing detection (which is coming)
* [x] added element scrollbar width: just outputting the width for now
   - e.g. users could set userContent/scripts to set the `scrollbar-width` to `none`, but this only affects the viewport, not elements
   * [ ] maybe moving the os detection logic to apply to the element measurement, rather than the viewport
* [x] Tor Browser for Android: find some `isTB` detection point: yay! found one, see next point
* [x] isTB2 (string) runs once as an additional "chrome" item
   - that is: I use the get_chrome function since the code is already set up
   - it is `resource://torbutton-assets/aboutTor.css`
   - it works on all platforms, and yay! finally something on Android
   - in the actual get_resource function, I don't output until isTB2 is either "y" or "n", and I factor both the tor-watermark.png  and the css files into the hash (for tor), and can now output Tor Browser on Android

🔻 **math**
* [ ] build a matrix (ECMAScript 1st vs 6th) of all results now TB9 (ESR68) has landed, and re-check output logic
   * note: FF68 changed the output of `expm1(1)` on most (but not all) setups
   * 99% sure I have all results correct so far: 18 results total: 10 < 68, 8 in 68+
* [ ] more tests: create more Linux distro VMs

🔻 **language**

* [x] added `Intl.NumberFormat`
   * [x] added `scientific notation` (currently nightly only, not riding any trains)
   * [x] added `compact long` (currently nightly only, not riding any trains)
   * [ ] expand with more types (e.g. literal, fraction etc)
* [x] added `Intl.NumberFormat(BigInt)` (FF70+)
* [x] expanded `[formatToParts] Intl.NumberFormat` with more [types](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat/formatToParts)
   * [x] prettify
   * [x] add codeChar to distinguish similar chars and show invisibles e.g. "narrow no-break space" for "fr"
   * [x] trap (and output) undefined (e.g hungarian group) and errors
   * [ ] expand with more types (e.g. currency, literal, percentSign etc)
* [x] date/time/format hash - add FF version notation
   * ~~check FF version notation with `verNo` and colorize notation red/green~~
   * ^^ update: we only add FF version notation if both `spoof_english` + `rfp` are green. Since the hash changes between releases are due to feature detection that can't be controlled by prefs, then the FF notation **must** be correct
   * [x] only output FF version notation if `isFirefox` : i.e FF specific stuff is only for firefox, but RFP etc stuff is for all (it's good to see if other browsers have the same result)
* [ ] add: `Intl.NumberFormat(BigInt).formatToParts` (FF70+) if it makes sense
* [x] added currency test

march 8th
* [x] moved app language PoCs to extras page since they have been resolved
* [x] added `unit percent` to Intl.NumberFormat

🔻 **cookies & storage**
* [ ] appcache test: decide what to do: the storage is deprecated in [FF71+](https://bugzilla.mozilla.org/show_bug.cgi?id=1237782) but not the API, and TB blocks it anyway.
* [ ] finish the `tests to come`

march 8th
* [x] combined multiple outputs into single lines -> shorter cleaner section

🔻 **devices**
* [x] add speech recognition [1248897](https://bugzilla.mozilla.org/show_bug.cgi?id=1248897)
* [x] fix touch test & added -moz-touch-enabled
* [x] add plugins test
* [x] add mime types test
* [x] add pointer event
* [x] add matchMedia any-pointer, pointer
* [x] enumerate speech engines (& RFP notation)
* [x] enumerate mediaDevices (& RFP notation)
* [x] added css and matchMedia `any-hover` + `hover`
* [x] added RFP notation to hover and pointer
* [ ] finish pointer tests
* [ ] finish touch tests on android: `radiusX/Y`, `screenX/Y`, `clientX/Y`
* [ ] finish enumerating gamepads (nfi how to test it with no devices)
* [ ] finish enumerating vrdisplays (nfi how to test it with no devices)

🔻 **canvas**
* [x] WebGL readPixels: added `tb safer` and `tb standard` notation based on error entropy
* [ ] WebGL readPixels: expand/test on error entropy e.g. in VM's (which usually block the graphics driver and the error messages can differ)
* [ ] add offscreenCanvas (currently disabled)
* [ ] add other tests, maybe: see [CB tests](https://canvasblocker.kkapsner.de/test/) e.g. detection tests

march 8th
* [x] when canvas is blocked (input and/or readout) this now outputs the error entropy
   - thanks @kkapsner 💋 👍 🍻 
* [x] all canvas (not webgl) values rather than just the hashes now have RFP notation
   * mozGetAsFile takes into account when it was deprecated

🔻 **domrect**
* [x] rebuilt to make sure error notation is always output in all possible scenarios with:
  - both initial page `loading` and section `re-run` button
  - both `file:///` and `https`
  - pref flipping `privacy.file_unique_origin`, iframe src file not found/blocked, and blocking iframes (e.g. uBO)
   - toggling any of the above 3 items between section **re-runs** (i.e not a page refresh)
* [ ] make sure each clientrect type runs and reports - e.g when blocked by CB

🔻 **audio**
* [x] added RFP notation for sampleRate
* [x] moved outputLatency from a separate display into the details + hash
   * hash improved to include all key names
   * number of keys notated
   * this was initially separate since it was the only thing RFP covered, and for macs calculating the RFP value relies on a subsequent key+value (sampleRate)
   * I now build an array, store the sampleRate value for later, and in the build the pretty detailed output including RFP notation from the array at the end
* [x] manual audio tests: audioContext test is now run after oscillator test
   - when run first: if RFP = false, the test will always return outputLatency = 0, which is incorrect, and a second test would run later to correct it
   - by running second: it is almost always correct and re-running the audioContext is not needed = saves time. It can still happen (I can sometimes trigger it on a cold Firefox, first run) - but I do detect and re-run again
* [x] manual audio tests: instead of timers to delay sequential runs: I use each test to start the next
   - oscillator -> context -> hybrid (-> context if required: to correct outputLatency = 0)
* [ ] add other tests, maybe: see [CB tests](https://canvasblocker.kkapsner.de/test/audioTest.html) e.g. different Hz

🔻 **fonts**
* [x] In FF69+ if the woff2 test fails, I return "blocked" (instead of "disabled [or blocked]") - the pref was removed in FF69
* [x] added a default font basic measurement test: alpha-numeric A-Z,a-z,0-9
   - this is similar to unicode glyphs: I render large single characters and measure them
   - this should pick up different default fonts (e.g Times New Roman on windows, Deja Vu Serif on Linux, Arimo on TB on Linux) as well as changes per font version, and graphics rendering
   - needs some data: I got two different hashes on Windows 7 bare metal vs windows 7 VM
* [ ] unicode glyphs: add more domrect measurements
* [ ] font code: fallback font first run: investigate further
   * when document fonts are blocked, the first run misses fonts it picks up on subsequent tests: currently mitigated by doing a 3-font "priming" run: but this is not an elegant solution
* [ ] font list: build the lists: win, mac, android, linux - see #43
* [ ] font list: cleanup monsta list - see #43
* [ ] test: default font family name: serif, sans-serif, monospace
* [ ] test: detect selected font versions: Times New Roman, Arial [thru unicode changes per version]

march 8th
* [x] fallback font test on TZP and extras page now uses the same test strings
   -  arthur's fallback font test uses his own list of codepoints, but I append all the ones from the unicode glyphs and the fpjs2 string: this actually improves the detection rate (with no false positives) - but on the extras page, the unicode glyphs code never runs and the appended codepoints never happen. This is now fixed
* [x] fixed the fallback font test string where two of the spans were a bit fucked up
* [x] unicode glyphs: added measureText
* [x] unicode glyphs: added `getClientRects` - this only shows if the hash differs from `getBoundingClientRect`
* [x] unicode glyphs: all three(measure, bounding, clientrects) indicate if they are blocked (e.g by CB)
* [x] unicode glyphs: domrect results append the number of unique differences out of unique measurements from measureText: the diffs are also output to console with logExtra

🔻 **media**

we'll get to this one day

🔻 **webgl**
* [ ] run tests for `webgl1` and `webgl2` separately: and break into logical sections for hashing and details
* [ ] make sure error entropy is handled for TB slider and VMs

🔻 **css**
* [x] `prefers-color-scheme` and `prefers-reduced-motion`: 2 tests -> 4 tests
   - previously used css to color hidden objects, then js to read the color back
   - now as pure css, and as pure JS (matchMedia with RFP notation)

🔻 **misc**
* [x] add beacon
* [x] add mediaSession
* [x] add web share
* [x] add component interfaces
* [x] add svg
* [x] add webauthn / u2f
* [x] add slider notation to mathml ~~and wasm~~ [edit wasm is only in alpha for now]
* [x] add navigator prototype (object.keys)
* [x] add reporting api
* [x] mathml: now zoom-resistent and more robust to changes in css line height
   - `mathmltest` element offsetHeight vs a non-mathml one
   - it used to check they were the same - now uses a range
* [ ] add timing rounding precision

🔻 **randomizing entropy**

detect randomization: not needed at this stage for `FF+RFP` or `TB` but nice to have
* [ ] audio
* [ ] canvas
* [ ] domrect
* [ ] clientrect in css line height
* [ ] clientrect in unicode glyphs

🔻 **other pages**
* [ ] timing - e.g. [this temp PoC](https://thorin-oakenpants.github.io/testing/timing.html)
* [ ] zombie persistent storage - e.g. [temp PoC](https://ghacksuserjs.github.io/TorZillaPrint/sanitizing.html)
* [ ] widgets / form controls: see [this](http://stephenhorlander.com/form-controls.html)
* [ ] elements

march 8th
* [x] added a tor browser chrome:// resource:// test

🔻 **to investigate**

* AOM (accessability object model)
* SVG: enabled/disabled, but also specifically look at
   * SVGTextContentElement.prototype.getComputedTextLength
   * SVGPathElement.prototype.getTotalLength
   * SVGGeometryElement.getPointAtLength()

...

**Comment by Thorin-Oakenpants** [comment]
I noticed domrect wouldn't run if iframes were blocked: and by that I mean I didn't pick up on it and output the error notation. I thought this was already solved ages ago. Anyway, I rebuilt the whole thing so it creates and destroys the iframe **every**.**single**.**time**, especially with all the problematic issues with the re-run button (i.e not a page refresh)

Here's a wee story
   * on a page refresh, everything worked fine, it was the re-run button
   * on a re-run the iframe src was already loaded: which caused issues
   * there are up to 5 possible situations for domrect to pan out
      - I do a lot of testing in file:///
      - 1. using file:///` the iframe src could be file not found
      - 2. using file:/// the pref `privacy.file_unique_origin` could be set to true (added default true in 68 AFAICRemeber) which causes a cross-origin resource sharing (cors) error
      - 3. using file:///` it could work
      - 4. on https:  the iframe src could be not found and/or read
      - 5. on https:  it could work
   * Major issues with `d`: e.g. blocking iframes (uBlockOrigin) and loading the page failed to output anything (no errors)
   * And to make matters worse, toggling uBOs iframe block state: as soon as you were blocking it once, you couldn't unblock on a re-rerun, and if you tried, on subsequent blocking, you ended up running the code twice (kinda).
   * What a mess
   * Long story short: On every load, on every re-rerun, in all scenarios, in all possible order of testing pref flips, renaming the source file between re-runs (to simulate file-not-found), uBO iframe toggling .. and in FF60-72 ... I now create a brand new iframe, load the source, and regardless of the outcome, I output the right message or results, and always successfully remove the listener and iframe for the next re-run.
   * Took me f^^^king ages
   * 🍺 🍻 

If you use https://thorin-oakenpants.github.io/testing/ where my prototype is working (while it stays there), you can test it all out. Open if from https, or copy the repo and run it as file:///. In file:///, feel free to pref flip privacy.file_unique_origin and re-run and flip and rerun etc, and rename the file to something else and re-run etc. And then on https, feel free to uBO toggle iframes on and off etc. Every single time now it works.

@kkapsner : do you like it? Test that sucker out!

**Comment by Thorin-Oakenpants** [comment]
Hey @kkapsner ... [1588980](https://bugzilla.mozilla.org/show_bug.cgi?id=1588980) .. comment 1 is priceless :)  "deprecated since FF26" ... the patch hasn't landed in my Nightly yet, but I assume the canvas code will output `not supported` for `mozGetAsFile` ?

**Comment by kkapsner** [comment]
I did not know that "and will be removed soon" is more than 6 years... ;)

Yes - it should output `not supported`.

**Comment by Thorin-Oakenpants** [comment]
> I did not know that "and will be removed soon" is more than 6 years... ;)

Back then they didn't have `Intl.RelativeTimeFormat` 🤣

**Comment by Thorin-Oakenpants** [comment]
@kkapsner I have a note here about `measureText` and `getLineDash` - i have no idea why I wrote them down or where I saw something that peaked my interest in them while perusing in my travels, but it would have been for a reason

Do you use them in your tests? Do you cover them in CB? Can they be exploited to detect spoofing?

Edit: just scoped out [getLineDash](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getLineDash) - that doesn't seem very revealing unless I'm missing something. Same with [measureTest](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/measureText)

I find the precision in the [w3schools example](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml5_canvas_measuretext) to possibly be revealing: if I change the width to 400 so I can see the full result, I get `154.48333740234375`

I think I answered my own Q as to why I made a note (at least with measureText) as it seems RFP doesn't block it (because - and thus its high precision could be telling - similar to unicode glyph measurments

**Comment by kkapsner** [comment]
I do not use them in my tests and CB does not cover them. I already touched measureText a little bit. Not sure why I stopped looking at it... but also not quite sure how to protect it properly.

I do not think it can be used to detect spoofing from CB but I think it can be used to do font enumeration.

**Comment by Thorin-Oakenpants** [comment]
> I do not think it can be used to detect spoofing from CB

Well, that's only because you're not spoofing it. Loading a new iframe an remeasuring the same object should be able to detect if the value is randomized in any way. And yeah, I'm think this is only useful for font measuring

Still not sure why `getLineDash` is in my ToDo list

**Comment by Thorin-Oakenpants** [comment]
not live, yet - I'll dig into it a bit more later: for now, I'm using the the unicode character list and appending the canvas text `measureText().width` for the `default` style only

The measureText return the same value as getBoundingClientRect - which is what I use for the clientrect hash: which is, of course, all six styles and width & height <- but you randomize those (although I have a way to somewhat de-randomize them).

Not terribly exciting, since I can't see how measureText can return height, but still useful. I'll probably expand the measureText to include all six styles, and remove it from the display output

![measureText](https://user-images.githubusercontent.com/16656956/75874349-6f1b5400-5e09-11ea-98c4-0911a036cd07.png)

**Comment by Thorin-Oakenpants** [comment]
OK, so here's a diff: 43 lines/chars, each with six measurements: measureText (width) on the left vs getBoundingClientRect (width) on the right. Some chars in some styles are different: seems measureText has higher precision

`U+2581`
- 1875.9832763671875 x 6 (measureText)  <-- higher precision
- 1875.9833984375 x 6 (cllentrect)

`U+05C6` has three different measurements
- measureText
   - 609.3833618164062  <-- higher precision
   - 705.0833129882812
   - 1200.199951171875
- clientrect
   - 609.38330078125
   - 705.0833129882812
   - 1200.199951171875

![measure-vs-cr](https://user-images.githubusercontent.com/16656956/75881422-f28f7200-5e16-11ea-8f16-e300b61088dd.png)

**Comment by Thorin-Oakenpants** [comment]
^^ actually, that's **not** more precision, it's "different"
e.g. if  `609.3833`0078125  was to be more precise, you wouldn't expect `609.3833`618164062

If it was going to be more precise you wouldn't no change expect the last one or two digits
e.g.  `609.383300781`25  -> something like `609.383300781`2475

**Comment by Thorin-Oakenpants** [comment]
using a small set of chars which have no difference between the two methods, it should now be possible to always detect when CB is randomizing clientrect

**Comment by KOLANICH** [comment]
>The measureText return the same value as getBoundingClientRect

That's a very good thing. In 2015 `measureText` have been returning floats.

But now the worst thing.
I have just tested my 2015 script ... it turned ot the result doesn't depend on:
* CanvasBlocker status (enabled/dissabled) (set to constant mode)
* RFP status.

The files with fonts dimensions are exactly the same! It means that neither Mozilla nor CB protect this API. It's a shame, since Brave code protecting the API is free and open source.

**Comment by Thorin-Oakenpants** [comment]
> The measureText return the same value as getBoundingClientRect

except I was wrong! I have asked someone at Mozilla how these two can have different results. I don't think it adds anything new to entropy or makes it worse.

RFP has plans/tickets to do *something* about clientrect but no one knows what yet: same probably goes for this - which is not as straight forward as it seems, because it's actually a font issue as well.

By that, I mean, I could specify the font be Times New Roman (rather than default) and check fonts are allowed to be set, etc, and that Times New Roman was actually loaded and used - to try and reduce the FPing entropy down to rendering/float differences or whatever they are. I could go further and use a downloadable font so that different versions of TNR on different OSes+updates don't create any differences.

Can you tell me where you found that measureText uses fractions / floats? A little over my head, but I understand floating points a little <-- this is probably the cause of the difference and you've nailed it. Math & floating points is how I detect all that math entropy: architecture, browser build architecture, os <-- [531915](https://bugzilla.mozilla.org/show_bug.cgi?id=531915)

**Comment by KOLANICH** [comment]
>Can you tell me where you found that measureText uses fractions / floats?

In fact I have missed it (that's why I don't use it), it's tor folks who have noticed it, when I have shown them my fingerprinter. Unfortunately I have missed their notice and rounded the dimensions to make them fit into u8 typed array in order to hash them. And had to stay that way because the fingerprints sent to me by people were made this way. It is important to preserve them.

BTW

**Comment by Thorin-Oakenpants** [comment]
OK, half of that makes no sense to me :) I'm active on trac.tor and I know of your ticket (well, there's one somewhere in there with KOLANICH stamped all over it) - I have edit access and try to make sure all the `tbb-fingerprinting*` keywords are added - I'm guessing that [13400](https://trac.torproject.org/projects/tor/ticket/13400) is it ?

It's a very old ticket with nothing new for 5 years. The PoCs are dead AFAIK. Its a very confusing ticket. I've tried to read you dull captaincy thing a couple of times before, but gave up (probably due to my lack of knowledge and partly due to English not being your native language).

I might contact dcf when I upload my next batch of changes so at least the measureText on those chars is live to look at. I also output a diff of unique changes to the console when logExtra = true

e.g.
```
[measureText vs getBoundingClientRect] 5 differences
1 1092.7667236328125 vs 1092.7666015625
2 1333.9832763671875 vs 1333.9833984375
3 1405.2667236328125 vs 1405.2666015625
4 1875.9832763671875 vs 1875.9833984375
5 609.3833618164062 vs 609.38330078125
```

I'll probably explore this more down the track when I get time: e.g using a downloadable and testing on different platforms

**Comment by Thorin-Oakenpants** [comment]
OK, so https://trac.torproject.org/projects/tor/attachment/ticket/13021/measureTextFP.html is a different kind of test

I already have made a note that I can use measureText to detect the default font: two assumptions
 - documents are not blocked
- canvas API is not blocked

Both those two assumptions should hold for Tor Browser. But loading and measuring all the fonts possible listed in TB's whitelisting, and then comparing it to the measurement of the default font, I should now know your default font - however, this is pointless in Tor Bowser, since your font is set by your locale/language and if you chose to spoof as en-US (on non en-US builds), your default font is changed to that of en-US. But by expanding the list to default fonts for Firefox, it could be effective in FF for picking up entropy on those who fuck around with their default fonts :)

So back to that test linked above (which I haven't tried), I don't think it has been (fully) mitigated (and I would use a much larger font size) as different versions of fonts and different platforms can still render differently - and then there's the question of if the way measureText is calculated (math) is involved. My test should be enough for this

But we'd really need separate tests to work out how much entropy is caused by each factor (font version, rendering, math)

**Comment by Thorin-Oakenpants** [comment]
aww man .. derailed my own thread .. nevermind, I have about 40 to 50 changes/new-things coming up in my next update (maybe next week), and I'll update the first post after that and then start a new one

anyway .. from the expert at Mozilla

> Yeah, getBoundingClientRect() uses app-units internally, which are the fixed-point representation we use for layout (1/60th of a CSS pixel), and thus have less precision than measureText, which uses floats.

**Comment by kkapsner** [comment]
@KOLANICH CB should protect getBoundingClientRect (but only changes maximal 0.005 - otherwise too many web pages break) but it does not touch measureText (but I think it should...). Where can I have a look at your test?
I had a very quick look into the Brave code and it seems that they simply do not return any measurements if the fingerprinting is not allowed. This is no option for CB as it would make it very easy detectable.

@Thorin-Oakenpants: values in an iframe should be the same as the parent window (see e.g. https://canvasblocker.kkapsner.de/test/test.html or https://canvasblocker.kkapsner.de/test/iframeTest.html).
When I will start tackling measureText I can make it always the same value the one from DOMRect if the original values are the same. So this shouldn't be a concern.

But the whole font list enumerating thing is out of scope for CB. It's too much built in to be tackled by a WebExtension. I just do not have the APIs to do it properly and without any wholes. I think the only way is to ship Firefox with a set of fonts and only use these. So no system fonts. And also do all the font computation in CPU and nothing in GPU...

PS: the 1/60th of a pixel is kind of odd as the numbers do not quite match. The decimal expansion of these numbers should then either be finite, have infinite 3s or infinite 6s at the end...

**Comment by Thorin-Oakenpants** [comment]
> values in an iframe should be the same as the parent window

when I last tested (I had a prototype locally where I could check if audio was being randomized) - all it took was to create and load a new iframe and the iframe called it's own js, which ran its results, and then the parent would check for and compare the results - this was enough to detect randomizing - maybe you changed how consistently you spoof per 1st party

But I dropped all the randomizing detection, as it more than often doubles the time taken to get that entropy, often way more if I have to create an iframe and wait for it to load. It's on my list but not important, especially since Tor Browser isn't affected ... but.... it's often not just a case of entropy being 1 bit (yes or no), since you provide so many options to toggle what to spoof - so the possible outcome exponential (not factorial, not permutations, some sort of combination) - I think I got that right (I am not a mathematician) e.g
- It's 2 (binary true/false) to the power of n <-- no idea what you call that
- 2 things you spoof -> 2^2 = 4 possibilities  (true+true, true+false, false+false, false+true)
- 3 things you spoof -> 2^3 = 8 possibilities
- 8 things you spoof -> 256

Granted, most people wouldn't change from your default settings, but I see 9 options in canvas/webgl, 7 options in audio, 8 options in domrect = up to 24 options. Not saying they can all be detected - but 24 binary entropy items is a lot of possible outcomes

whether it actually has any real world entropy or is worth FPing (time resources vs payoff) is questionable

PS: you didn't bite when I said I could de-randomize somewhat your clientrect spoofing - I didn't look at it further, since I dropped the randomizing detection work: but I had a workaround to hash the decimal places you **never** randomize from what I could tell in my tests (and it wasn't just the first 3 decimal places or something: there was a pattern, at least on my machine) - I didn't explore it further

**Comment by KOLANICH** [comment]
>I'm guessing that 13400 is it ?

Not quite. https://trac.torproject.org/projects/tor/ticket/14310

And there is no need to guess: it is explicitly said in https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting/blob/master/readme.en.tex#L231

>It's a very old ticket with nothing new for 5 years.

Yeah.

>The PoCs are dead AFAIK.

The actual PoCs are in my GH or GL account https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting/blob/master/fingerprinter.html . Sometimes I check if this has been fixed, it hasn't.

Please notice that generating a PDF in GitLab CI doesn't work for now.

>I've tried to read you dull captaincy thing a couple of times before, but gave up (probably due to my lack of knowledge and partly due to English not being your native language).

I guess it is style. The original text was a blog post on a Russian website similar to medium.com (but predating it) and contained some memes, some of which are not global. When translating it to English I decided to keep the style.

@kkapsner

> Where can I have a look at your test?

https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting

the fingerprinter is in `fingerprint.js` .

> CB should protect getBoundingClientRect

My fingerprinter [combines `getBoundingClientRect` and `measureText` into a single hash](https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting/blob/master/fingerprinter.js#L212L215).

>(but only changes maximal 0.005 - otherwise too many web pages break)

BTW, what do you think about https://bugzilla.mozilla.org/show_bug.cgi?id=1450564 ?

**Comment by Thorin-Oakenpants** [comment]
> Not quite. https://trac.torproject.org/projects/tor/ticket/14310

Never seen it before. 5 years old, no activity, and no keywording. It's also a kind of generic tickets with no specifics

With FPing, it's important to evaluate the damage each "item" can do, rather than a mixed bag: e.g. canvas font fingerprinting is two things: canvas and fonts. Same with combining getBoundingClientRect and measureText

It's fine to produce FPing with multiple aspects to show that it can be done - I mean that's what FPing actually is: combined measurements/metrics => an overall unique hash. for example: the dom rect tests combine fonts and widgets etc -> https://ghacksuserjs.github.io/TorZillaPrint/iframes/domrect.html

I'm still learning .. but there's a lot of overlap in tests: math/floats + dpi + devicePixelRatio and even default fonts vs font versions of the same font: it muddies the waters when trying to work out how to tackle the issue

---

remember this?
> Yeah, getBoundingClientRect() uses app-units internally, which are the fixed-point representation we use for layout (1/60th of a CSS pixel), and thus have less precision than measureText, which uses floats.

so I asked "Is a "CSS pixel" affected by subpixels - e.g. devicePixelRatio > 1? " and got this reply

> It's not, in theory, but text layout goes from device pixels (which _are_ affected by DPI) to CSS pixels, so in practice yeah, there's dpi-dependent stuff going on in there.

So it's entirely feasible that measureText is affected by devicePixelRatio : and to test that we would have to eliminate all other variables

**Comment by Thorin-Oakenpants** [comment]
So this is slightly interesting (and to be expected I guess)

Since CB doesn't affect measureText, it could perhaps be inferred that spoofing is being applied to that particular clientrect when the differences exceed a certain threshold. There are 43 characters with 6 styles = 258 measurements

Pic: at the top there is no getBoundingClientRect spoofing, at the bottom there is is
![diffs](https://user-images.githubusercontent.com/16656956/76051597-342d3380-5f63-11ea-912e-577e3bdaad69.png)

Those figures mean 5 unique changes out of 20 different values

What that % threshold is, IDK. And it should really be calculated as unique changes out of unique values (I think I'll add that) - in other words: when you spoof, the chances of another difference increases. But who knows, maybe 20 or so unique diffs is common on some devices

And I can easily expand the test to include more than just getBoundingClientRect, so individual settings could be inferred.

anti-FPing should be always be implemented carefully, otherwise it can add entropy: whether that new entropy outweighs the old entropy is relevant, but leaking new entropy should be avoided if possible
 - e.g: when Tor Browser added a maximize warning panel, it was (and still is) possible to infer the application language by measuring the panel height (the application language is used for the message, and the text/font causes differences). Is it better to **a)** warn people (i.e maximized potentially leaks more entropy than app language) or **b)** not warn (app language is less entropy) - or **c)** find a solution that informs the end user but does not alter the viewport height. I know what I would do - **option c**.
- e.g. letterboxing's implementation also provides some new entropy, but the benefits far outweigh those

**Comment by Thorin-Oakenpants** [comment]
^^ and this particular test (inferring various clientrect spoofing) could always be improved with a better set of characters for consistent results

**Comment by Thorin-Oakenpants** [comment]
ok, so I added unique values returned by measureText (in my case it's 31)
- no spoof:  5 / 31 = different = `~16%`
- spoof: 22 / 31 = different = `~71%`

I also played with CB's `DOM Rectinteger factor`
- 0: big surprise: back to 5 diffs
- 1: 26 diffs
- 2: 22 diffs
- 4 (default): 22 diffs
- 7: 22 diffs
- 10: 24 diffs
- 13: 26 diffs
- 60: 26 diffs

**Comment by kkapsner** [comment]
@KOLANICH:
> My fingerprinter [combines `getBoundingClientRect` and `measureText` into a single hash](https://github.com/KOLANICH/Article-2015-Dull-captaincy-or-the-way-Tor-Project-fights-browser-fingerprinting/blob/master/fingerprinter.js#L212L215).

But the `measurement` is a `Int8Array` so the CB changes do not survive.

> BTW, what do you think about https://bugzilla.mozilla.org/show_bug.cgi?id=1450564 ?

I think it's a good approach. I actually would also make web pages look more consistent across different devices/OSes.

@Thorin-Oakenpants:
> maybe you changed how consistently you spoof per 1st party

There was always a setting to spoof consistent per domain. But it's not set by default (it's the persistent random number generator).
Also I changed the iFrame behaviour some versions ago so that code in the iFrame should return the same hashes (my tests actually do not run code in the iFrame itself - but it's on my todo list).

> Granted, most people wouldn't change from your default settings, but I see 9 options in canvas/webgl, 7 options in audio, 8 options in domrect = up to 24 options. Not saying they can all be detected - but 24 binary entropy items is a lot of possible outcomes

Actually these options should not be touched unless there are some major website breakages. If they are used without the understanding what the actual functions do the browser could easily be detected.

Since CB uses a different approach against fingerprinting (increase entropy and change it all the time) it's hard to talk about increasing the trackable entropy. But you are right: fingerprint protection is not a trivial task. And the next version of CB will therefore protect measureText (https://github.com/kkapsner/CanvasBlocker/issues/448).

> you didn't bite when I said I could de-randomize somewhat your clientrect spoofing

I do not take every bait ;-)... will have to look at it.

**Comment by Thorin-Oakenpants** [comment]
> I do not take every bait ;-)... will have to look at it.

Well, I'm not baiting you :) I was just surprised you didn't follow up on it. This was maybe 6 months ago - but there was definitely a pattern (I did not change the dom rect integer however): namely that only certain decimal places were ever altered, consistently - leaving the rest as stable, so I could always use those decimal places in the fingerprint. I haven't ever had the time to follow up on it.

> Actually these options should not be touched unless there are some major website breakages. If they are used without the understanding what the actual functions do the browser could easily be detected.

When you added those options per API, it became great for testing things - but I could already see that they would be problematic. I think they should be hidden and defaults used unless the user agrees they are an advanced user? <-- probably too much work. Stick a yellow warning symbol ⚠️ with an info/tooltip in each of the three API sections if they are not at default <-- that's what I would do, probably. Not really thought about it.

**Comment by kkapsner** [comment]
> Well, I'm not baiting you :) I was just surprised you didn't follow up on it.

I just overlooked it...

> I think they should be hidden and defaults used unless the user agrees they are an advanced user?

They are behind the "expert mode" flag.

> Stick a yellow warning symbol ⚠️ with an info/tooltip in each of the three API sections if they are not at default

I could add a warning in the settings sanitation.

**Comment by KOLANICH** [comment]
>But the measurement is a Int8Array so the CB changes do not survive.

Anyway, the hashes are pretty unique. Interestingly, on ReactOS fingerprints don't depend on fonts files installed (I have removed the remappings)  and RFP mode (in fact I haven't seen the newly installed (by just copying) fonts in WordPad, I haven't tried installing via an `Install` button because I don't want to install them manually).

**Comment by kkapsner** [comment]
> Anyway, the hashes are pretty unique.

That's the problem with font fingerprinting. And as it's a so fundamental problem I cannot solve it with CB. If would try to fake the values on the integer level it would be immediately detectable. And I think "pretty unique" is still more than then 60 000 users of CB. So I have to stay under the radar.

**Comment by Thorin-Oakenpants** [comment]
@kkapsner 

> > Can you give me your test page where you saw that increasing the factor increases the number of differences between DOMRect and TextMetrics

It's build directly into my local copy of TZP, which has about 20 other changes, not to mention lots of cleaner code, better variable names etc - and I'm waiting on several more changes before I do a bulk update. It's all a bit messy trying to separate out my changes just for this

I'll see what I can do, but here's the basics of it

```css
.offscreen {position: absolute; top: -2000%;}
```

```html
<div class="offscreen">
	<div id="ugDiv" class="normalized" style='font-size: 22pt;'>
		<span id="ugSpan"><span id="ugSlot"></span></span>
	</div>
</div>
```

```js
var fntCode = ['0x20B9','0x2581','0x20BA','0xA73D','0xFFFD','0x20B8','0x05C6',
	'0x1E9E','0x097F','0xF003','0x1CDA','0x17DD','0x23AE','0x0D02','0x0B82','0x115A',
	'0x2425','0x302E','0xA830','0x2B06','0x21E4','0x20BD','0x2C7B','0x20B0','0xFBEE',
	'0xF810','0xFFFF','0x007F','0x10A0','0x1D790','0x0700','0x1950','0x3095','0x532D',
	'0x061C','0x20E3','0xFFF9','0x0218','0x058F','0x08E4','0x09B3','0x1C50','0x2619']

function get_str_from_codepoint(n) {
	if (n <= 0xffff) {
		return String.fromCharCode(n)
	} else {
		n -= 0x10000
		return String.fromCharCode(0xd800 + (n >> 10), 0xdc00 + (n % 0x400))
	}
}

var logExtra = true
var logPerf = false

function get_unicode() {
	/* code based on work by David Fifield (dcf) and Serge Egelman (2015)
		https://www.bamsoftware.com/talks/fc15-fontfp/fontfp.html#demo */
	// vars
	let styles = ["default","sans-serif","serif","monospace","cursive","fantasy"],
		div = dom.ugDiv, span = dom.ugSpan, slot = dom.ugSlot,
		offset = [], client = [], measure = [], munique = [], diffs = [], display = "",
		canvas = dom.ugCanvas, ctx = canvas.getContext("2d"),
		t0 = performance.now()
	// each char
	for (let i=0; i < fntCode.length; i++) {
		let n = fntCode[i],
			c = get_str_from_codepoint(n),
			cp = "u+" + n.substr(2)
		display += "<br>" + cp.padStart(7)
		// each style
		for (let j=0; j < styles.length; j++) {
			// set
			slot.style.fontFamily = (j == 0 ? "" : styles[j])
			slot.textContent = c
			// offset
			let w = span.offsetWidth, h = div.offsetHeight
			offset.push((j==0 ? cp+"-" : "" ) + w+"x"+h)
			display += (w.toString()).padStart(8) +" x "+ (h.toString()).padStart(4)
			// clientrect
			let eDiv = div.getBoundingClientRect()
			let eSpan = span.getBoundingClientRect()
			w = eSpan.width; h = eDiv.height
			client.push( (j==0 ? cp+"-" : "" ) + w+"x"+h )
			// measuretext
			try {
				ctx.font = "normal normal 22000px " + (j == 0 ? "none" : styles[j])
				let m = ctx.measureText(c).width
				measure.push( (j==0 ? cp+"-" : "" ) + m)
				munique.push(m)
				if (m !== w) {diffs.push(m+" vs "+w)}
			} catch(err) {}
		}
	}
	// clear
	dom.ugSlot = ""
	// diffs
	let total = diffs.length
	diffs = diffs.filter(function (item, position) {return diffs.indexOf(item) === position})
	munique = munique.filter(function (item, position) {return munique.indexOf(item) === position})
	if (logExtra) {
		console.log("glyphs: measuretext vs getbounding")
		for (let i=0; i < diffs.length; i++) {console.debug (i+1, diffs[i])}
	}
	// output
	dom.ug1 = sha1(offset.join())
	dom.ug2.innerHTML = sha1(measure.join())
		+ (measure.length == 0 ? sb+"[canvas blocked]"+sc : " ["+ diffs.length +"|"+ munique.length +"|"+ total +" diffs]")
	dom.ug3 = sha1(client.join())
	dom.ug4.innerHTML = fntHead + display
	// perf
	if (logPerf) {debug_log("unicode glyphs [fonts]",t0)}
}
```

**Comment by Thorin-Oakenpants** [comment]
^^ you can skip the output part

- diffs is an array of where `measureText` didn't match `getBounding*` width: I reduce that by removing duplicates
   - so `diffs.length` = unique diffs (e.g. 5 or 24 etc)
- munique is just an array of the measureText widths, which again, I reduce by removing duplicates
   - so `munique.length` = unique values (e.g. 31 in my case)

the diffs array contains the values

![diffs](https://user-images.githubusercontent.com/16656956/76135611-55078e80-6020-11ea-8784-a58bc641e149.png)


I also plan to investigate adding more than just getBoundingClientRect: just add more measurement methods and arrays etc - and compare each DOMect hash to each other (they should be the same)

Is that enough for you to go on? Because, honestly, I'm a week or two away from my next update for TZP

**Comment by Thorin-Oakenpants** [comment]
> > Anyway, the hashes are pretty unique.
> 
> That's the problem with font fingerprinting. And as it's a so fundamental problem I cannot solve it with CB. If would try to fake the values on the integer level it would be immediately detectable. And I think "pretty unique" is still more than then 60 000 users of CB. So I have to stay under the radar.

Fundamentally, the question is should it be defended, and how to defend: and each type should be evaluated on its on
- e.g spoofing UA is the stupidest thing in the world - it does not work and can be bypassed (at least in FF) and is only good for compat reasons. Defending this does more harm that good (unless in an enforced set such as RFP where everyone does it). I do get that
- e.g. spoofing canvas - if you do nothing you are highly FP'able - spoofing or blocking or returning a set value like RFP = detectable or not, is better than nothing. in RFP's case, everyone is the same in an enforced set, and detection doesn't matter

IDK why Tor Browser insists on lowering entropy instead of randomizing. This is some randomizing in RFP (media stats is one), and I prefer simple set spoofs where possible. But randomizing is the proper and probably only solution in a lot of cases (domrect, audio - webgl maybe), and I feel like Tor Project are sticking their heads in the sand.

A lot of fingerprinters/scripts are dumb: FPing likes stability: It's easy AF to cause instability in these scripts - and this should be a weapon TB uses. I wish they would listen more to me :)

**Comment by kkapsner** [comment]
> Is that enough for you to go on?

Yes - thanks.

**Comment by KOLANICH** [comment]
>I feel like Tor Project are sticking their heads in the sand.

I feel like randomization is mostly needed only as an anti-discrimination measure. Some websites discriminate against users of software protecting their privacy (blacklist approach) or against users not using the software and hardware, violating their privacy (say thanks to TEE hardware manufacturers and developers for that, so the game is already lost in tech (one cannot pay millions for extracting a key from each chip sold), your only hope is to ban this kind of discrimination by making a law banning it explicitly and making a law legalizing the circumvention of this discrimination also explicitly, to prevent lawsuits in order to bully ones doing that).

So, we first strip all the useful information we can. And then we add some misinformation. Misinformation can be averaged out (it must be countered, but it is impossible to counter all the possible attacks), but if the useful info is stripped and not reintroduced, it cannot be inferred. So stripping first is very important.

**Comment by Thorin-Oakenpants** [comment]
I disagree: stripping should be the last option: when all other avenues are exhausted, or as a stop-gap measure

Whether something is used for bullying or not is immaterial: that's a side effect. The fact that it can increase entropy is the actual problem: therefore it needs to be dealt with - so you evaluate the possible solutions: of which I would say there are about eight options: all would have different pros and cons depending on the problem

In no particular order
- outright disabling the API / blocking
- spoofing all users the same
- spoofing in buckets (e.g. letterboxing steps)
- spoofing based on some criteria (e.g. by OS: such as audio ac-sampleRate)
- randomizing
- randomizing in buckets
- randomizing based on some criteria

Additionally (the eight option), one can minimize the risk by blocking the FP'ing scripts: e.g.
  - security slider = safest
  - adding Firefox's ETP fingerprinters (and hopefully over time it will become better)
  - incorporating something like google's privacy thing whatever the F its called: something along the lines of each request for some client side info carries a value: and after a website uses up it's quota, the end user is prompted or something
      * asking for screen res = 1 point
      * asking for canvas readout = 5 points
      * every 10 system fonts asked for = 2 points

The problem with randomizing in a lot of things, is that it creates information paradoxes and causes breakage. Hence e.g. randomizing by a set criteria (such as OS) can stop that breakage (and OS is easily determined: you'll never hide that).

But things like [most of: poo-poohs to UA spoofing] CB's stuff: webaudio, domrect, canvas, webgl = are pretty much universal, and I don't see any solution except just full on randomizing. There is a solution for webGL called unityGL or something: it removes all the floats or something and everyone on all browsers end up the same, and nothing breaks.

**Comment by Thorin-Oakenpants** [comment]
just updated OP - not sure if I covered everything: but the number of checks went up by about 30?
